podpac/core/utils.py

```diff
@@ -329,3 +329,41 @@ def _get_from_url(url):
     except RuntimeError as e:
         _log.warning("Cannot authenticate to {}. Check credentials. Error was as follows:".format(url) + str(e))
     return r.text
+
+
+def cached_property(fn):
+    """Decorator that creates a property that is cached as a private attribute."""
+
+    key = "_podpac_cached_property_%s" % fn.__name__
+
+    @property
+    def wrapper(self):
+        if hasattr(self, key):
+            value = getattr(self, key)
+        else:
+            value = fn(self)
+            setattr(self, key, value)
+        return value
+
+    return wrapper
+
+
+def cached_node_property(fn):
+    """Decorator that creates a property that is cached as a private attribute and in the Node cache_ctrl."""
+
+    key = "_podpac_cached_property_%s" % fn.__name__
+
+    @property
+    def wrapper(self):
+        if hasattr(self, key):
+            value = getattr(self, key)
+        elif self.has_cache(key):
+            value = self.get_cache(key)
+            setattr(self, key, value)
+        else:
+            value = fn(self)
+            setattr(self, key, value)
+            self.put_cache(value, key)
+        return value
+
+    return wrapper
```

Comment:
@mpu-creare I added two cached property decorators:
 * `cached_property`: store the property as a private attribute
 * `cached_node_property`: store the property as a private attribute *and* using the node `cache_ctrl`

What do you think? I used the `cached_property` everywhere I thought it made sense throughout podpac core. Most of those can be replaced with `cached_node_property` (but a special few cannot). So the question is: do you think I should use the podpac cache version `@cached_node_property` everywhere I can, instead? Another way to put this is: do you see `@cached_property` or `@cached_node_property` as the "common" case?

I will rename them according to your answer.

Reply:
Basically, I have an *internal* use for the one that doesn't use the podpac cache, but if you only want to provide users with the one that *does* use the podpac cache, I want to reduce this to one decorator and handle the other cases a special way.

But if you think we do want both, I'll leave it (though maybe you can suggest a better pair of names).

Reply:
I like the two decorators. You could probably also do it with an input to the decorator, but I really don't mind two. How about: `cached_property` and `cached_property_persistent` as the former only works for a session, while the latter *could* work across sessions depending on the user's cache preferences.